[main] Fix LoaderAllocator computation for a generic type instance

alexey-zakharov opened the pull request:
## Purpose of this PR
Improve logic of computing LoaderAllocator for a generic type instance to avoid AssemblyLoadContext leaks.
Resolves https://github.com/dotnet/runtime/issues/111611

**Problem:**
`ClassLoader::ComputeLoaderModuleWorker` uses a LoaderAllocator creation number to determine the latest LoaderAllocator to place a class in. However the method does not account the number of the type definition itself when calculating the latest number and this can lead into a situation when the generic instance is put into the older LoaderAllocator.

E.g. if we have type `ClassA` from AssemblyLoadContext `A` with number 2 and type `ClassB` from AssemblyLoadContext `B` with number 5 the resulting loader allocator for the type ClassB<ClassA> would correspond to AssemblyLoadContext `A`. If `ClassB` contains a static, such static would be then placed under the LoaderAllocator `A`, preventing AssemblyLoadContext `B` from unloading.

**Proposed solution:**
The PR updates the `ComputeLoaderModuleWorker` logic to account for the LoaderAllocator of a defining type, so we use it if it has the latest creation number.

(cherry picked from commit b3a73f3b66589df3cd8d00c7333ad28e84d669d2)

alexey-zakharov committed:
Take defining type loader allocator into account when computing loader allocator of a generic type instance

(cherry picked from commit b3a73f3b66589df3cd8d00c7333ad28e84d669d2)

alexey-zakharov commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -176,22 +179,17 @@ PTR_Module ClassLoader::ComputeLoaderModuleWorker(
             }
         }
 
         // Only if we didn't find a different loader allocator than the defining loader allocator do we
```
@joncham spotted, that perhaps this part of the comment is not needed anymore

alexey-zakharov commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -113,18 +113,15 @@ PTR_Module ClassLoader::ComputeLoaderModuleWorker(
-        pLoaderModule = pDefinitionModule;
-    }
+    // Use the definition module as the loader module by default
+    Module *pLoaderModule = pDefinitionModule;
```
if pDefinitionModule != null we effectively fall through to it if it is collectible or not at the end, so I've decided to remove the if block

alexey-zakharov commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -155,14 +154,18 @@ PTR_Module ClassLoader::ComputeLoaderModuleWorker(
+        if (pDefinitionModule != NULL)
+        {
+            pOldestLoaderModule = pDefinitionModule;
+            latestFoundNumber = pDefinitionModule->GetLoaderAllocator()->GetCreationNumber();
```
assumption here is that if module in non-collectible ALC (or Default it still has a LoaderAllocator), but perhaps if it is not collectible we need to ignore it in favor of a first collectible number 🤔

davidwrighton commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -176,22 +179,18 @@ PTR_Module ClassLoader::ComputeLoaderModuleWorker(
+            if (pLoaderAllocatorCheck->IsCollectible() &&
+                pLoaderAllocatorCheck->GetCreationNumber() > latestFoundNumber)
             {
                 pOldestLoaderModule = pModuleCheck;
```
Please rename `pOldestLoaderModule` to something like `pLatestLoaderModule` The name is a legacy of very long ago when I was doing experiments to build this feature, and ... is rather inaccurate.

alexey-zakharov replied:
done in 3a4e07cf8a8142c25f533eb28b53b908b84fa948

davidwrighton reviewed (changes requested):
As you note, a test would be helpful. However, since the behavior of this sort of computation is not actually specified and may change in the future without warning, I'd rather not merge it in with existing tests, and would prefer this to be a separate test (and for the test to have some comments in it indicating that failure is reasonable when running on different runtimes). I'm pretty flexible on naming though, as I don't have a strong opinion. I'm notoriously terrible at naming stuff.

alexey-zakharov committed:
Added a test that validates ALC unload when it participates in a static field of a generic class

alexey-zakharov committed:
Addressed codereview comment

alexey-zakharov committed:
Updated https://github.com/mono/mono/issues/15142 to https://github.com/dotnet/runtime/issues/34072

alexey-zakharov commented:
> would prefer this to be a separate test (and for the test to have some comments in it indicating that failure is reasonable when running on different runtimes)

I've looked at the best location for the test and found that there is `src/libraries/System.Runtime.Loader/tests/CollectibleAssemblyLoadContextTest.cs` suite which tests ALC unloadability and seems like a perfect fit for the issue - would it be acceptable to have a test there?

- added `Unload_TwoCollectibleWithOneAssemblyAndOneInstanceReferencingAnotherThroughGenericStatic` test which tests the failing scenario
- updated issue links since the mentioned issue is closed now

alexey-zakharov commented on src/libraries/System.Runtime.Loader/tests/AssemblyLoadContextTest.cs:
```diff
@@ -189,7 +185,7 @@ public static void DefaultAssemblyLoadContext_Properties()
             Assert.Contains("System.Runtime.Loader.DefaultAssemblyLoadContext", alc.ToString());
             Assert.Contains(alc, AssemblyLoadContext.All);
-            Assert.Contains(Assembly.GetCallingAssembly(), alc.Assemblies);
+            Assert.Contains(typeof(int).Assembly, alc.Assemblies);
```
I've updated the test because it was failing as the calling assembly for the test method is xunit assembly which is not in the Default ALC and I suppose is test runner implementation specific
CorLib assembly which is where type of int is is more stable check parameter imho

alexey-zakharov commented:
@davidwrighton I don't see any suspicious test failures - do you think the changes are good to go?

alexey-zakharov committed:
Merge branch 'main' into dotnet-upstream/fix-computeloaderallocator

alexey-zakharov committed:
Merge branch 'main' into dotnet-upstream/fix-computeloaderallocator

alexey-zakharov committed:
Merge branch 'main' into dotnet-upstream/fix-computeloaderallocator

alexey-zakharov committed:
Merge branch 'main' into dotnet-upstream/fix-computeloaderallocator
